File: api/src/main/java/com/coffeeshop/api/service/ClientService.java
package com.coffeeshop.api.service;

import com.coffeeshop.api.model.ClientEntity;
import com.coffeeshop.api.repository.ClientRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

@Service
@RequiredArgsConstructor
public class ClientService {

    private final ClientRepository clientRepository;

    public ClientEntity register(String name, String password) throws Exception {
        var client = clientRepository.getClientByName(name);
        if (client != null) throw new Exception("User already exists");
        return clientRepository.addClient(name, password);
    }

    public ClientEntity login(String name, String password) throws Exception {
        var client = clientRepository.checkPassword(name, password);
        if (client == null) throw new Exception("Invalid password");
        return client;
    }

    public Long getBonusPoints(Long clientId) throws Exception {
        var bonusPoints = clientRepository.getBonusPoints(clientId);
        if (bonusPoints == null) throw new Exception("User doesnt exists");
        return bonusPoints;
    }

    public void spendPoints(Long clientId, Long amount) {
        clientRepository.spendBonusPoints(Math.toIntExact(clientId), Math.toIntExact(amount));
    }

    public void saveFavouriteCoffee(Long clientId, Long beansId, Long milkId, Long syrupId, Long sizeId) {
        clientRepository.saveFavouriteCoffee(Math.toIntExact(clientId), Math.toIntExact(beansId), Math.toIntExact(milkId), Math.toIntExact(syrupId), Math.toIntExact(sizeId));
    }

}

File: front/src/app/services/client.service.ts
import {Injectable} from '@angular/core';
import {HttpClient} from "@angular/common/http";
import {ClientModel} from "../model/ClientModel";

@Injectable({
  providedIn: 'root'
})
export class ClientService {

  url = 'http://localhost:52520/api/client';

  user: ClientModel | null = null
  bonusPoints: number = 0

  constructor(private http: HttpClient) {
  }

  register(name: string, password: string) {

    return this.http.post<ClientModel>(this.url + "/registration", {
      name: name,
      password: password
    }, {
      withCredentials: true
    })

  }

  login(name: string, password: string) {

    return this.http.post<ClientModel>(this.url + "/login", {
      name: name,
      password: password
    }, {
      withCredentials: true
    })

  }

  saveFavouriteCoffee(clientId: number, milkId: number, beansId: number, syrupId: number, sizeId: number) {

    return this.http.post(this.url + "/favourite-coffee", {
      client_id: clientId,
      milk_id: milkId,
      beans_id: beansId,
      syrup_id: syrupId,
      size_id: sizeId
    }, {
      withCredentials: true
    })

  }

  getBonusPoints(clientId: number) {
    return this.http.get<number>(this.url + `/bonus-points?id=${clientId}`, {
      withCredentials: true
    })
  }

  spendBonusPoints(clientId: number, bonusPoints: number) {

    return this.http.post(this.url + "/bonus-points", {
      client_id: clientId,
      bonus_points: bonusPoints
    }, {
      withCredentials: true
    })

  }

}
